### PR TITLE
More robust exe_exist function

### PR DESCRIPTION
**Pull request by MadsAndreasen**
The documentation for subprocess.run at https://docs.python.org/3/library/subprocess.html#popen-constructor has a warning, pointing to using shutil.which() instead of subprocess.run for detecting if exe files exists on the path.

I have experienced problems with detecting llvm-cov, when running as non-root in a container. In this situation subprocess.run would throw a Permission denied exception if llvm-cov is not on the path.

### CODE REVIEW

**Comment by eli-schwartz** [comment]
The commit message is definitely broken in and of itself. It says nothing at all, whereas the GitHub web UI has significantly more detail that is mandatory for future readers of the code to understand why this change was made.

In terms of the actual code change, if exe_exists is a wrapper function that immediately calls shutil.which with no other function body code, then it would make more sense to delete the function and, everywhere it is used, replace it with direct use of shutil.which. But it seems strange to get Permission Denied, anyway...

**Comment by MadsAndreasen** [comment]
> The commit message is definitely broken in and of itself. It says nothing at all, whereas the GitHub web UI has significantly more detail that is mandatory for future readers of the code to understand why this change was made.
> 
> In terms of the actual code change, if exe_exists is a wrapper function that immediately calls shutil.which with no other function body code, then it would make more sense to delete the function and, everywhere it is used, replace it with direct use of shutil.which. But it seems strange to get Permission Denied, anyway...

It is definitely strange, and I haven't been able to figure out why.
Anyway, I'd be happy to remove the function call and use which() instead. 
And of course put some more info in the commit.

**Comment by eli-schwartz** [comment]
From what I can tell, the logic to call the program with --version dates back all the way to c3c9a31a5a51ea2b2c7b560262a5af0d372d52ab which contained inline versions of that logic.

```
git grep exe_exists
```

6 hits including the function definition.

```
git grep shutil.which
```

43 hits, none of which are the definition since it is in the stdlib.

I wonder if exe_exists simply exists because no one ever thought to change it.

**Comment by bruchar1** [comment]
There is a small typo in the commit message: `shutil.whichk()`

**Comment by jpakkane** [comment]
FWICT `exe_exists` is now called only from unit tests. If that is the case then let's just delete the whole thing.

> I wonder if exe_exists simply exists because no one ever thought to change it.

Quite possible. Or that this was working around some bug that existed 10 years ago but no longer does.

File: mesonbuild/environment.py [hunk 1]
```diff
@@ -105,13 +105,13 @@ def detect_llvm_cov(suffix: T.Optional[str] = None):
             tool = 'llvm-cov'
         else:
             tool = f'llvm-cov-{suffix}'
-        if mesonlib.exe_exists([tool, '--version']):
+        if shutil.which(tool) is not None:
```
Comment (eli-schwartz) [inline mesonbuild/environment.py:108]: You can just do `if shutil.which(tool)` and in other cases "if not shutil.which(....)" -- the return value of shutil is always a positive string, or None if it wasn't found.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: mesonbuild/utils/universal.py [hunk 2]
```diff
@@ -683,14 +682,8 @@ def is_qnx() -> bool:
...
-def exe_exists(arglist: T.List[str]) -> bool:
-    try:
-        if subprocess.run(arglist, timeout=10).returncode == 0:
-            return True
-    except (FileNotFoundError, subprocess.TimeoutExpired):
-        pass
-    return False
-
+def exe_exists(exe: str) -> bool:
+    return shutil.which(exe) is not None
```
Comment (eli-schwartz) [inline mesonbuild/utils/universal.py:686]: I thought we were removing this entirely?

File: unittests/internaltests.py [hunk 3]
```diff
@@ -813,6 +814,10 @@ def test_msvc_toolset_version(self):
         self.assertTrue(vctools_ver.startswith(toolset_ver),
                         msg=f'{vctools_ver!r} does not start with {toolset_ver!r}')
 
+    def test_exe_exists(self):
+        self.assertTrue(utils.universal.exe_exists('python3'))
+        self.assertFalse(utils.universal.exe_exists('command_that_does_not_exist'))
```
Comment (eli-schwartz) [inline unittests/internaltests.py:819]: What does this test?

File: mesonbuild/utils/universal.py (hunk 2 above)
Comment (MadsAndreasen) [reply mesonbuild/utils/universal.py:686]: Created https://github.com/mesonbuild/meson/pull/13456 to remove the function entirely.

**Pull request merged**